cmd/syncthing/upnpsvc.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) 2015 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/syncthing/syncthing/internal/config"
+	"github.com/syncthing/syncthing/internal/upnp"
+)
+
+// The UPnP service runs a loop for discovery of IGDs (Internet Gateway
+// Devices) and setup/renewal of a port mapping.
+type upnpSvc struct {
+	cfg       *config.Wrapper
+	localPort int
+}
+
+func newUPnPSvc(cfg *config.Wrapper, localPort int) *upnpSvc {
+	return &upnpSvc{
+		cfg:       cfg,
+		localPort: localPort,
+	}
+}
+
+func (s *upnpSvc) Serve() {
+	extPort := 0
+	noIGDFound := false
+
+	for {
+		igds := upnp.Discover(time.Duration(s.cfg.Options().UPnPTimeoutS) * time.Second)
+		if len(igds) >= 0 {
+			noIGDFound = false
+			extPort = s.tryIGDs(igds, extPort)
+		} else if !noIGDFound {
+			noIGDFound = true
+			l.Infof("No UPnP device detected")
+		}
+
+		d := time.Duration(s.cfg.Options().UPnPRenewalM) * time.Minute
+		if d == 0 {
+			// We always want to do renewal so lets just pick a nice sane number.
+			d = 30 * time.Minute
+		}
+		time.Sleep(d)
+	}
+}
+
+func (s *upnpSvc) Stop() {
+	panic("upnpSvc cannot stop")
+}
+
+func (s *upnpSvc) tryIGDs(igds []upnp.IGD, prevExtPort int) int {
+	// Lets try all the IGDs we found and use the first one that works.
+	// TODO: Use all of them, and sort out the resulting mess to the
+	// discovery announcement code...
+	for _, igd := range igds {
+		extPort, err := s.tryIGD(igd, prevExtPort)
+		if err == nil {
+			if extPort != prevExtPort {
+				// External port changed; refresh the discovery announcement.
+				// TODO: Don't reach out to some magic global here?
+				l.Infof("New UPnP port mapping: external port %d to local port %d.", extPort, s.localPort)
+				discoverer.StopGlobal()
+				discoverer.StartGlobal(s.cfg.Options().GlobalAnnServers, uint16(extPort))
+			}
+			if debugNet {
+				l.Debugf("Created/updated UPnP port mapping for external port %d on device %s.", extPort, igd.FriendlyIdentifier())
+			}
+			return extPort
+		} else {
+			l.Warnf("Failed to set UPnP port mapping: external port %d on device %s.", extPort, igd.FriendlyIdentifier())
+		}
+	}
+
+	return 0
+}
+
+func (s *upnpSvc) tryIGD(igd upnp.IGD, suggestedPort int) (int, error) {
+	leaseTime := s.cfg.Options().UPnPLeaseM * 60
+
+	if suggestedPort != 0 {
+		// First try renewing our existing mapping.
+		name := fmt.Sprintf("syncthing-%d", suggestedPort)
+		err := igd.AddPortMapping(upnp.TCP, suggestedPort, s.localPort, name, leaseTime)
+		if err == nil {
+			return suggestedPort, nil
+		}
+	}
+
+	var err error
```

Comment:
Should move this to the top of the function, and make `err :=` into `err =` above.